sdks/python/apache_beam/transforms/util.py

```diff
@@ -485,33 +485,57 @@ class ReshufflePerKey(PTransform):
   """
 
   def expand(self, pcoll):
-    class ReifyTimestamps(DoFn):
-      def process(self, element, timestamp=DoFn.TimestampParam):
-        yield element[0], TimestampedValue(element[1], timestamp)
+    windowing_saved = pcoll.windowing
+    if windowing_saved.is_default():
+      # In this (common) case we can use a trivial trigger driver
+      # and avoid the (expensive) window param.
+      _globally_windowed = window.GlobalWindows.windowed_value(None)
+      window_fn = window.GlobalWindows()
+      MIN_TIMESTAMP = window.MIN_TIMESTAMP
+
+      def reify_timestamps(element, timestamp=DoFn.TimestampParam):
+        key, value = element
+        if timestamp == MIN_TIMESTAMP:
+          timestamp = None
+        return key, (value, timestamp)
+
+      def restore_timestamps(element):
+        key, values = element
+        return [
+            _globally_windowed.with_value((key, value))
+            if timestamp is None
+            else window.GlobalWindows.windowed_value((key, value), timestamp)
+            for (value, timestamp) in values]
+
+    else:
+      # The linter is confused.
+      # pylint: disable=abstract-class-instantiated
```

Comment:
PyCharm says this class is missing `to_runner_api_parameter`.

Reply:
PyCharm is confused too. It inherits this via the urn magic.